{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Wed Apr 19 15:14:28 2017\n",
    "\n",
    "@author: vrkrishn\n",
    "\"\"\"\n",
    "import sys\n",
    "\n",
    "usage = sys.argv[0] + \": genome.fasta features.gff\"\n",
    "\n",
    "if len(sys.argv) < 3:\n",
    "    print(usage)\n",
    "    sys.exit(\"\\nThis script requires both a FASTA file and a GFF file\\n\")\n",
    "\n",
    "genome = sys.argv[1]\n",
    "gff    = sys.argv[2]\n",
    "\n",
    "\n",
    "# Read FASTA file\n",
    "data1 = open(genome, 'r')\n",
    "\n",
    "# Variable that will hold the genome sequence\n",
    "gen = ''\n",
    "\n",
    "# Initialize a line counter\n",
    "line_num = 0\n",
    "\n",
    "# Read in the genome file\n",
    "for line in data1:\n",
    "    line = line.rstrip('\\n')\n",
    "\n",
    "    if line_num > 0:\n",
    "        gen = gen + line\n",
    "\n",
    "    # increment line_number\n",
    "    line_num += 1\n",
    "\n",
    "# Close the genome file\n",
    "data1.close()\n",
    "\n",
    "# Read GFF file\n",
    "data2 = open(gff, 'r')\n",
    "\n",
    "# Variables for all the feature types \n",
    "CDS     = ''\n",
    "tRNA    = ''\n",
    "rRNA    = ''\n",
    "IN      = ''\n",
    "MF      = ''\n",
    "RR      = ''\n",
    "\n",
    "# read in the GFF file\n",
    "for line in data2:\n",
    "\n",
    "    # remove newline's - could also use strip\n",
    "    line = line.rstrip('\\n')\n",
    "\n",
    "    types = line.split('; type ')\n",
    "    other_type = types[len(types)-1]\n",
    "    # print(other_type)\n",
    "    \n",
    "    cols = line.split('\\t')\n",
    "    type  = cols[2]\n",
    "    start = int(cols[3])\n",
    "    end   = int(cols[4])\n",
    "    \n",
    "    # print(type, \"\\t\", start, \"\\t\", end)\n",
    "\n",
    "    # extract this feature from the genome\n",
    "    fragment = gen[start-1:end]\n",
    "\n",
    "    if type == 'CDS':\n",
    "        CDS += fragment\n",
    "\n",
    "    if type == 'intron':\n",
    "        IN += fragment\n",
    "\n",
    "    if type == 'misc_feature':\n",
    "        MF += fragment\n",
    "\n",
    "    if type == 'repeat_region':\n",
    "        RR += fragment\n",
    "\n",
    "    if type == 'rRNA':\n",
    "        rRNA += fragment\n",
    "\n",
    "    if type == 'tRNA':\n",
    "        tRNA += fragment\n",
    "\n",
    "# Length of feature types and genome sequence\n",
    "gen1 = len(gen)\n",
    "exon = len(CDS)\n",
    "intron = len(IN)\n",
    "misc = len(MF)\n",
    "repeat = len(RR)\n",
    "trna = len(tRNA)\n",
    "rrna = len(rRNA)\n",
    "\n",
    "# GC Content\n",
    "GC_CDS = round((CDS.count('G') + CDS.count('C'))/exon*100, 2)\n",
    "GC_RR  = round((RR.count('G') + RR.count('C'))/repeat*100, 2)\n",
    "GC_MF  = round((MF.count('G') + MF.count('C'))/misc*100, 2)\n",
    "GC_IN  = round((IN.count('G') + IN.count('C'))/intron*100, 2)\n",
    "GC_tRNA  = round((tRNA.count('G') + tRNA.count('C'))/trna*100, 2)\n",
    "GC_rRNA  = round((rRNA.count('G') + rRNA.count('C'))/rrna*100, 2)\n",
    "\n",
    "\n",
    "# Percent of the genome covered by featured type\n",
    "exon1 = round(exon/gen1*100, 2)\n",
    "intron1 = round(intron/gen1*100, 2)\n",
    "misc1 = round(misc/gen1*100, 2)\n",
    "repeat1 = round(repeat/gen1*100, 2)\n",
    "tr = round(trna/gen1*100, 2)\n",
    "rr = round(rrna/gen1*100, 2)\n",
    "\n",
    "# Print the output\n",
    "print('CDS', exon, str(exon1) + '%', str(GC_CDS))\n",
    "print('Intron', intron, str(intron1) + '%', str(GC_IN))\n",
    "print('Misc Features', misc, str(misc1) + '%', str(GC_MF))\n",
    "print('Repeat Region', repeat, str(repeat1) + '%', str(GC_RR))\n",
    "print('tRNA', trna, str(tr) + '%', str(GC_tRNA))\n",
    "print('rRNA', rrna, str(rr) + '%', str(GC_rRNA))\n",
    "\n",
    "# close the GFF file\n",
    "data2.close()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
